automerge-c/src/lib.rs

```diff
@@ -0,0 +1,363 @@
+        Some(AMresult::Values(_)) => AmStatus::ValuesOk,
+        Some(AMresult::Changes(_)) => AmStatus::ChangesOk,
+        None => AmStatus::InvalidResult,
+    }
+}
+
+/// \memberof AMdoc
+/// \brief Set a map object's value.
+///
+/// \param[in] doc A pointer to an `AMdoc` struct.
+/// \param[in] obj A pointer to an `AMobj` struct.
+/// \param[in] key A map object's key string.
+/// \param[in] data_type An `AmDataType` enum tag matching the actual type that
+///            \p value points to.
+/// \param[in] value A pointer to the value at \p key or `NULL`.
+/// \return A pointer to an `AMresult` struct containing no value.
+/// \pre \p doc must be a valid address.
+/// \pre \p obj must be a valid address.
+/// \pre \p key must be a valid address.
+/// \warning To avoid a memory leak, the returned pointer must be deallocated
+///          with `AMclear()`.
+///
+/// # Safety
+/// doc must be a pointer to a valid AMdoc
+/// obj must be a pointer to a valid AMobj or NULL
+/// key must be a c string of the map key to be used
+/// value must be a pointer to data of the type specified in data_type
+#[no_mangle]
+pub unsafe extern "C" fn AMmapSet(
+    doc: *mut AMdoc,
+    obj: *mut AMobj,
+    key: *const c_char,
+    data_type: AmDataType,
+    value: *const c_void,
+) -> *mut AMresult {
+    let doc = to_doc!(doc);
+    to_result!(doc.set(to_obj!(obj), to_prop!(key), to_value!(value, data_type)))
+}
+
+/// \memberof AMdoc
+/// \brief Set a list object's value.
+///
+/// \param[in] doc A pointer to an `AMdoc` struct.
+/// \param[in] obj A pointer to an `AMobj` struct.
+/// \param[in] index A list object's index number.
+/// \param[in] data_type An `AmDataType` enum tag matching the actual type that
+///            \p value points to.
+/// \param[in] value A pointer to the value at \p index or `NULL`.
+/// \return A pointer to an `AMresult` struct containing no value.
+/// \pre \p doc must be a valid address.
+/// \pre \p obj must be a valid address.
+/// \warning To avoid a memory leak, the returned pointer must be deallocated
+///          with `AMclear()`.
+/// # Safety
+/// doc must be a pointer to a valid AMdoc
+/// obj must be a pointer to a valid AMobj or NULL
+/// value must be a pointer to data of the type specified in data_type
+#[no_mangle]
+pub unsafe extern "C" fn AMlistSet(
+    doc: *mut AMdoc,
+    obj: *mut AMobj,
+    index: usize,
+    data_type: AmDataType,
+    value: *const c_void,
+) -> *mut AMresult {
+    let doc = to_doc!(doc);
+    to_result!(doc.set(to_obj!(obj), index, to_value!(value, data_type)))
+}
+
+/// \memberof AMresult
+/// \brief Get an `AMresult` struct's `AMobj` struct value.
+///
+/// \param[in] result A pointer to an `AMresult` struct.
+/// \return A pointer to an `AMobj` struct.
+/// \pre \p result must be a valid address.
+///
+/// # Safety
+/// result must be a pointer to a valid AMresult
+#[no_mangle]
+pub unsafe extern "C" fn AMgetObj(_result: *mut AMresult) -> *mut AMobj {
+    unimplemented!()
```

Comment:
As above, is this intended to be implemented soon or should we document that it's unimplemented?

Reply:
> As above, is this intended to be implemented soon...?

@alexjg, yes. I'm trying to expose everything.

Reply:
> As above, is this intended to be implemented soon or should we document that it's unimplemented?

@alexjg, the `AMgetObj()` function is supplanted by the `AMgetValue()` function in PR #310.